File: apollo-runtime/src/main/java/com/apollographql/apollo/internal/cache/normalized/NoOpApolloStore.java
package com.apollographql.apollo.internal.cache.normalized;

import com.apollographql.apollo.api.GraphqlFragment;
import com.apollographql.apollo.api.Operation;
import com.apollographql.apollo.api.Response;
import com.apollographql.apollo.api.ResponseFieldMapper;
import com.apollographql.apollo.cache.CacheHeaders;
import com.apollographql.apollo.cache.normalized.ApolloStore;
import com.apollographql.apollo.cache.normalized.ApolloStoreOperation;
import com.apollographql.apollo.cache.normalized.CacheKey;
import com.apollographql.apollo.cache.normalized.CacheKeyResolver;
import com.apollographql.apollo.cache.normalized.NormalizedCache;
import com.apollographql.apollo.cache.normalized.Record;

import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.UUID;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * An alternative to {@link RealApolloStore} for when a no-operation cache is needed.
 */
public final class NoOpApolloStore implements ApolloStore, ReadableStore, WriteableStore {

  @Override public Set<String> merge(@NotNull Collection<Record> recordCollection, @NotNull CacheHeaders cacheHeaders) {
    return Collections.emptySet();
  }

  @Override public Set<String> merge(Record record, @NotNull CacheHeaders cacheHeaders) {
    return Collections.emptySet();
  }

  @Nullable @Override public Record read(@NotNull String key, @NotNull CacheHeaders cacheHeaders) {
    return null;
  }

  @Override public Collection<Record> read(@NotNull Collection<String> keys, @NotNull CacheHeaders cacheHeaders) {
    return Collections.emptySet();
  }

  @Override public void subscribe(RecordChangeSubscriber subscriber) {
  }

  @Override public void unsubscribe(RecordChangeSubscriber subscriber) {
  }

  @Override public void publish(Set<String> keys) {
  }

  @NotNull @Override public ApolloStoreOperation<Boolean> clearAll() {
    return ApolloStoreOperation.emptyOperation(Boolean.FALSE);
  }

  @NotNull @Override public ApolloStoreOperation<Boolean> remove(@NotNull CacheKey cacheKey) {
    return ApolloStoreOperation.emptyOperation(Boolean.FALSE);
  }

  @NotNull @Override public ApolloStoreOperation<Integer> remove(@NotNull List<CacheKey> cacheKeys) {
    return ApolloStoreOperation.emptyOperation(0);
  }

  @Override public ResponseNormalizer<Map<String, Object>> networkResponseNormalizer() {
    //noinspection unchecked
    return ResponseNormalizer.NO_OP_NORMALIZER;
  }

  @Override public ResponseNormalizer<Record> cacheResponseNormalizer() {
    //noinspection unchecked
    return ResponseNormalizer.NO_OP_NORMALIZER;
  }

  @Override public <R> R readTransaction(Transaction<ReadableStore, R> transaction) {
    return transaction.execute(this);
  }

  @Override public <R> R writeTransaction(Transaction<WriteableStore, R> transaction) {
    return transaction.execute(this);
  }

  @Override public NormalizedCache normalizedCache() {
    return null;
  }

  @Override public CacheKeyResolver cacheKeyResolver() {
    return null;
  }

  @NotNull @Override
  public <D extends Operation.Data, T, V extends Operation.Variables> ApolloStoreOperation<T> read(
      @NotNull Operation<D, T, V> operation) {
    return ApolloStoreOperation.emptyOperation(null);
  }

  @NotNull @Override
  public <D extends Operation.Data, T, V extends Operation.Variables> ApolloStoreOperation<Response<T>> read(
      @NotNull Operation<D, T, V> operation, @NotNull ResponseFieldMapper<D> responseFieldMapper,
      @NotNull ResponseNormalizer<Record> responseNormalizer, @NotNull CacheHeaders cacheHeaders) {
    return ApolloStoreOperation.emptyOperation(Response.<T>builder(operation).build());
  }

  @NotNull @Override
  public <F extends GraphqlFragment> ApolloStoreOperation<F> read(@NotNull ResponseFieldMapper<F> fieldMapper,
      @NotNull CacheKey cacheKey, @NotNull Operation.Variables variables) {
    return ApolloStoreOperation.emptyOperation(null);
  }

  @NotNull @Override
  public <D extends Operation.Data, T, V extends Operation.Variables> ApolloStoreOperation<Set<String>> write(
      @NotNull Operation<D, T, V> operation, @NotNull D operationData) {
    return ApolloStoreOperation.emptyOperation(Collections.<String>emptySet());
  }

  @NotNull @Override
  public <D extends Operation.Data, T, V extends Operation.Variables> ApolloStoreOperation<Boolean> writeAndPublish(
      @NotNull Operation<D, T, V> operation, @NotNull D operationData) {
    return ApolloStoreOperation.emptyOperation(Boolean.FALSE);
  }

  @NotNull @Override
  public ApolloStoreOperation<Set<String>> write(@NotNull GraphqlFragment fragment, @NotNull CacheKey cacheKey,
      @NotNull Operation.Variables variables) {
    return ApolloStoreOperation.emptyOperation(Collections.<String>emptySet());
  }

  @NotNull @Override
  public ApolloStoreOperation<Boolean> writeAndPublish(@NotNull GraphqlFragment fragment, @NotNull CacheKey cacheKey,
      @NotNull Operation.Variables variables) {
    return ApolloStoreOperation.emptyOperation(Boolean.FALSE);
  }

  @NotNull @Override
  public <D extends Operation.Data, T, V extends Operation.Variables> ApolloStoreOperation<Set<String>>
  writeOptimisticUpdates(@NotNull Operation<D, T, V> operation, @NotNull D operationData, @NotNull UUID mutationId) {
    return ApolloStoreOperation.emptyOperation(Collections.<String>emptySet());
  }

  @NotNull @Override
  public <D extends Operation.Data, T, V extends Operation.Variables> ApolloStoreOperation<Boolean>
  writeOptimisticUpdatesAndPublish(@NotNull Operation<D, T, V> operation, @NotNull D operationData,
      @NotNull UUID mutationId) {
    return ApolloStoreOperation.emptyOperation(Boolean.FALSE);
  }

  @NotNull @Override
  public ApolloStoreOperation<Boolean> rollbackOptimisticUpdatesAndPublish(@NotNull UUID mutationId) {
    return ApolloStoreOperation.emptyOperation(Boolean.FALSE);
  }

  @NotNull @Override public ApolloStoreOperation<Set<String>> rollbackOptimisticUpdates(@NotNull UUID mutationId) {
    return ApolloStoreOperation.emptyOperation(Collections.<String>emptySet());
  }
}

File: apollo-normalized-cache/src/jvmMain/kotlin/com/apollographql/apollo/cache/normalized/internal/NoOpApolloStore.kt
package com.apollographql.apollo.cache.normalized.internal

import com.apollographql.apollo.api.Fragment
import com.apollographql.apollo.api.Operation
import com.apollographql.apollo.api.Response
import com.apollographql.apollo.api.Response.Companion.builder
import com.apollographql.apollo.api.internal.json.JsonReader
import com.apollographql.apollo.cache.CacheHeaders
import com.apollographql.apollo.cache.normalized.ApolloStore
import com.apollographql.apollo.cache.normalized.ApolloStoreOperation
import com.apollographql.apollo.cache.normalized.ApolloStoreOperation.Companion.emptyOperation
import com.apollographql.apollo.cache.normalized.CacheKey
import com.apollographql.apollo.cache.normalized.CacheKeyResolver
import com.apollographql.apollo.cache.normalized.NormalizedCache
import com.apollographql.apollo.cache.normalized.Record
import java.util.UUID

/**
 * An alternative to RealApolloStore for when a no-operation cache is needed.
 */
class NoOpApolloStore : ApolloStore, ReadableStore, WriteableStore {
  override fun merge(recordCollection: Collection<Record>, cacheHeaders: CacheHeaders): Set<String> {
    return emptySet()
  }

  override fun merge(record: Record, cacheHeaders: CacheHeaders): Set<String> {
    return emptySet()
  }

  override fun stream(key: String, cacheHeaders: CacheHeaders): JsonReader? {
    return null
  }

  override fun read(key: String, cacheHeaders: CacheHeaders): Record? {
    return null
  }

  override fun read(keys: Collection<String>, cacheHeaders: CacheHeaders): Collection<Record> {
    return emptySet()
  }

  override fun subscribe(subscriber: ApolloStore.RecordChangeSubscriber) {}
  override fun unsubscribe(subscriber: ApolloStore.RecordChangeSubscriber) {}
  override fun publish(keys: Set<String>) {}
  override fun clearAll(): ApolloStoreOperation<Boolean> {
    return emptyOperation(java.lang.Boolean.FALSE)
  }

  override fun remove(cacheKey: CacheKey, cascade: Boolean): ApolloStoreOperation<Boolean> {
    return emptyOperation(java.lang.Boolean.FALSE)
  }

  override fun remove(cacheKey: CacheKey): ApolloStoreOperation<Boolean> {
    return emptyOperation(java.lang.Boolean.FALSE)
  }

  override fun remove(cacheKeys: List<CacheKey>): ApolloStoreOperation<Int> {
    return emptyOperation(0)
  }

  override fun <R> readTransaction(transaction: Transaction<ReadableStore, R>): R {
    return transaction.execute(this)!!
  }

  override fun <R> writeTransaction(transaction: Transaction<WriteableStore, R>): R {
    return transaction.execute(this)!!
  }

  override fun normalizedCache(): NormalizedCache {
    error("Cannot get normalizedCache: no cache configured")
  }

  override fun cacheKeyResolver(): CacheKeyResolver {
    return CacheKeyResolver.DEFAULT
  }

  override fun <D : Operation.Data> readOperation(
      operation: Operation<D>): ApolloStoreOperation<D> {
    error("Cannot read operation: no cache configured")
  }

  override fun <D : Operation.Data> readOperationInternal(
      operation: Operation<D>,
      cacheHeaders: CacheHeaders
  ): ApolloStoreOperation<Response<D>> {
    // This is called in the default path when no cache is configured, do not trigger an error
    // Instead return an empty response. This will be seen as a cache MISS and the request will go to the network.
    return emptyOperation(builder<D>(operation).build())
  }

  override fun <D : Fragment.Data> readFragment(
      fragment: Fragment<D>,
      cacheKey: CacheKey,
  ): ApolloStoreOperation<D> {
    error("Cannot read fragment: no cache configured")
  }

  override fun <D : Operation.Data> writeOperation(
      operation: Operation<D>,
      operationData: D,
      publish: Boolean
  ): ApolloStoreOperation<Set<String>> {
    // Should we throw here instead?
    return emptyOperation(emptySet())
  }

  override fun <D: Fragment.Data> writeFragment(
      fragment: Fragment<D>,
      cacheKey: CacheKey,
      fragmentData: D,
      publish: Boolean
  ): ApolloStoreOperation<Set<String>> {
    // Should we throw here instead?
    return emptyOperation(emptySet())
  }



  override fun <D : Operation.Data> writeOptimisticUpdates(operation: Operation<D>, operationData: D, mutationId: UUID): ApolloStoreOperation<Set<String>> {
    // Should we throw here instead?
    return emptyOperation(emptySet())
  }

  override fun <D : Operation.Data> writeOptimisticUpdatesAndPublish(operation: Operation<D>, operationData: D,
                                                                     mutationId: UUID): ApolloStoreOperation<Boolean> {
    // Should we throw here instead?
    return emptyOperation(java.lang.Boolean.FALSE)
  }

  override fun rollbackOptimisticUpdatesAndPublish(mutationId: UUID): ApolloStoreOperation<Boolean> {
    // Should we throw here instead?
    return emptyOperation(java.lang.Boolean.FALSE)
  }

  override fun rollbackOptimisticUpdates(mutationId: UUID): ApolloStoreOperation<Set<String>> {
    // Should we throw here instead?
    return emptyOperation(emptySet())
  }
}
